### ISSUE TITLE
Deprecation warnings cause IE8 errors when `process.env.NODE_ENV !== 'production'`

### ISSUE THREAD

**Issue by jesstelford** [facebook/react#3188]
**Version**: occurs in `0.13.0-beta.1` and `0.13.0-beta.2` (I'm not sure about earlier versions)

**Bug**:
For example, [this code](https://github.com/facebook/react/blob/d59a039ebdebebf5c98ab96026529f66063979bf/src/modern/class/ReactComponent.js#L106-L121):

```javascript
  var defineDeprecationWarning = function(methodName, displayName) {
    try {
      Object.defineProperty(ReactComponent.prototype, methodName, {
        get: function() {
          warning(
            false,
            '%s(...) is deprecated in plain JavaScript React classes.',
            displayName
          );
          return undefined;
        }
      });
    } catch (x) {
      // IE will fail on defineProperty (es5-shim/sham too)
    }
  };
```

Will cause `Exception thrown and not caught` errors in IE8 (with [es5-shim/sham v4.1.0](https://github.com/es-shims/es5-shim/tree/v4.1.0) installed) *regardless of if you call the deprecated methods or not*.

This is due to es5-sham [only supporting shimming of `value`](https://github.com/es-shims/es5-shim/blob/v4.1.0/es5-sham.js#L326) (not `get` as used above) property descriptors:

```javascript
    Object.defineProperty = function defineProperty(object, property, descriptor) {
        // ...
        // If it's a data property.
        if ('value' in descriptor) {
            // fail silently if 'writable', 'enumerable', or 'configurable'
            // are requested but not supported
            // ...
        } else {
            if (!supportsAccessors) {
                throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
            }
            // ...
        }
        return object;
    };
```

**Workaround**
Set `NODE_ENV='production'`. This disabled debug messages, and hence the deprecation warnings.

**Comment by zpao** [comment]
cc @sebmarkbage (and @syranide for IE8 expertise)

We have a try/catch there, so even if es5sham is throwing an error, we should be catching it there (thus the comment in the catch - we expect that behavior). You're sure it's coming from this block?

**Comment by jesstelford** [comment]
N...no. I can't reproduce the bug anymore :/

That was only one example of where `defineProperty` is used, maybe there's another instance that isn't wrapped in a try/catch. Let me see if I can reproduce the issue again.

**Comment by jesstelford** [comment]
Ok, got it reproduced now. It only happens in `0.13.0-beta.1` - works fine in `0.13.0-beta.2` :+1:

**Comment by syranide** [comment]
Fixed by #3017 :)
